lib/puppet/datatypes.rb

```diff
@@ -0,0 +1,190 @@
+# and passing it a block that defines the data type interface and implementation.
+#
+# Data types are namespaced inside the modules that contains them. The name of the
+# data type is prefixed with the name of the module. As with all type names, each
+# segment of the name must start with an uppercase letter.
+#
+# @example A simple data type
+#   Puppet::DataTypes.create_type('Auth::User') do
+#     interface <<-PUPPET
+#       attributes => {
+#         name => String,
+#         email => String
+#       }
+#     PUPPET
+#   end
+#
+# The above example does not declare an implementation which makes it exactly similar
+# to adding the following contents in a file named 'user.pp' under the 'types' directory
+# of the module root.
+#
+#   type Auth::User = Object[
+#     attributes => {
+#       name => String,
+#       email => String
+#     }]
+#
+# Both declarations are valid and will be found by the module loader.
+#
+# Anatomy of a data type
+# ---
+#
+# Data types consist of an interface and an implementation. A data type can often be used
+# by only declaring the interface. Unless an implementation is found it will be automatically
+# generated.
+#
+# In some cases, the generated implementation is not enough. In other cases it might be
+# necessary to declare a data type that maps to an already existing implementation.
+#
+# @example Adding implementation on top of the generated type using `implementation`
+#   Puppet::DataTypes.create_type('Auth::User') do
+#     interface <<-PUPPET
+#       attributes => {
+#         name => String,
+#         email => String,
+#         year_of_birth => Integer,
+#         age => { type => Integer, kind => derived }
+#       }
+#       PUPPET
+#
+#     implementation do
+#       def age
+#         DateTime.now.year - @year_of_birth
+#       end
+#     end
+#   end
+#
+# @example Appointing an already existing implementation class
+#
+# Assumes the following class is declared under 'lib/auth/utils' in the module:
+#
+#   class Auth::Utils::User
+#     attr_reader :name, :year_of_birth
+#     def initialize(name, year_of_birth)
+#       @name = name
+#       @year_of_birth = year_of_birth
+#     end
+#
+#     def age
+#       DateTime.now.year - @year_of_birth
+#     end
+#
+#     def send_text(sender, text)
+#       sender.send_text_from(@name, text)
+#     end
+#   end
+#
+# Then the type declaration can look like this:
+#
+#   Puppet::DataTypes.create_type('Auth::User') do
+#     interface <<-PUPPET
+#       attributes => {
+#         name => String,
+#         email => String,
+#         year_of_birth => Integer,
+#         age => { type => Integer, kind => derived }
+#       },
+#       functions => {
+#         send_text => Callable[Sender, String[1]]
+#       }
+#       PUPPET
+#
+#     # This require is optional and only needed in case
+#     # the implementation is not loaded by other means.
+#     require 'auth/utils/user.rb'
+#
+#     implementation_class Auth::Utils::User
```

Comment:
Shouldn't the name of the class be a String so you can load the Puppet Data Type without having to also load the implementation (say you only need the interface)?

Reply:
I guess we could introduce that if we think there's a need for it. The current `PObjectType` type doesn't contain any logic for deferred loading based on strings. The use of `implementation_class` is intended for pre-existing Ruby classes and should really be avoided for other purposes.